pyflagser/flagser.py

```diff
@@ -51,87 +164,88 @@ def flagser(flag_matrix, min_dimension=0, max_dimension=np.inf, directed=True,
-        computation. A negative value computes highest possible precision.
+        computation. If ``None``, no approximation is made and all cells are
+        used. For more details, please refer to [1]_.
 
     Returns
     -------
     out : dict of list
-        A dictionary holding the results of the flagser computation. Each
-        value is a list of length `max_dimension` - `min_dimension` + 1. The
-        key-value pairs in `out` are as follows:
+        A dictionary with the following key-value pairs:
 
         - ``'dgms'``: list of ndarray of shape ``(n_pairs, 2)``
           A list of persistence diagrams, one for each dimension greater
           than or equal than `min_dimension` and less than `max_dimension`.
           Each diagram is an ndarray of size (n_pairs, 2) with the first
           column representing the birth time and the second column
           representing the death time of each pair.
-        - ``'cell_count'``: list of int
-          Cell count per dimension greater than or equal than
-          `min_dimension` and less than `max_dimension`.
         - ``'betti'``: list of int
-          Betti number per dimension greater than or equal than
-          `min_dimension` and less than `max_dimension`.
-        - ``'euler'``: list of int
-          Euler characteristic per dimension greater than or equal than
+          Betti number at filtration value `max_edge_weight` per dimension
+          greater than or equal than `min_dimension` and less than
+          `max_dimension`.
+        - ``'cell_count'``: list of int
+          Cell count (number of simplices) at filtration value
+          `max_edge_weight` per dimension greater than or equal than
           `min_dimension` and less than `max_dimension`.
+        - ``'euler'``: int
+          Euler characteristic at filtration value `max_edge_weight`.
 
     Notes
     -----
-    For more details, please refer to the `flagser documentation \
-    <https://github.com/luetge/flagser/blob/master/docs/\
-    documentation_flagser.pdf>`_.
-
-    """
-    vertices = np.asarray(flag_matrix.diagonal()).copy()
-
-    if not approximation:
-        approximation = -1
+    The input graphs cannot contain self-loops, i.e. edges that start and end
+    in the same vertex, therefore diagonal elements of the input adjacency
+    matrix stores vertex weights.
 
-    if isinstance(flag_matrix, np.ndarray):
-        row = np.indices(flag_matrix.shape)[0].flat
-        column = np.indices(flag_matrix.shape)[1].flat
-        data = flag_matrix.flat
-        mask_off_diag = np.logical_not(np.eye(vertices.shape[0],
-                                              dtype=bool).flat)
+    References
+    ----------
+    .. [1] D. Luetgehetmann, "Documentation of the C++ flagser library";
+           `GitHub: <https://github.com/luetge/flagser/blob/master/docs/\
+           documentation_flagser.pdf>`_.
 
+    """
+    # Handle default parameters
+    if max_edge_weight is None:
```

Comment:
I don't know if I understand why we still have to look for the maximum value allowed by the `dtype`. If `_max_edge_weight` is the maximum value allowed by the `dtype` then by definition no filtering can be done using it. I.e. line 67 in `pyflagser._utils.py` does nothing useful. So I think nothing would change if this entire if-else block disappeared, and we fed `max_edge_weight` directly to `extract_weighted_graph` below.

Reply:
In the current implementation if `max_edge_weight` is ``None``, then np.inf when `dtype` is ``np.float`` and the maximum integer when `dtype` is ``np.int`` will be filtered. This was done simply because ``np.inf`` should always be filtered so it is never passed to C++ flagser.

Reply:
@gtauzin thanks but I still don't understand. The only relevant line in `extract_weighted_graph` seems to me to be:
```
    # Infinite weights mask
    if max_edge_weight is not None:
        mask = np.logical_and(mask, data <= max_edge_weight)
```
Here `max_edge_weight` is assigned the value `_max_edge_weight` which, as you said, is `np.inf` when the `dtype` is float and the largest available integer otherwise. Taking the float example for simplicity, it seems to me that
```
        mask = np.logical_and(mask, data <= max_edge_weight)
```
or
```
        mask = np.logical_and(mask, data)
```
would have the same effect. So why not just pass the user-selected `max_edge_weight` as is, and then modifying the line in `extract_weighted_graph` above to e.g.
```
    # Infinite weights mask
    if max_edge_weight is not None:
        mask = np.logical_and(mask, data <= max_edge_weight)
    else:
    mask = np.logical_and(mask, data)
```
?

Reply:
I am not sure I understand if data has a ``0`` entry talking it directly as a mask will not work as it will become an unwanted ``False``.

I think the solution is just to replace the ``<=`` by ``<``, no?

Reply:
> I am not sure I understand if data has a 0 entry talking it directly as a mask will not work as it will become an unwanted False

Oh yes indeed!  I missed that subtlety.  I like the idea of using <=, so how about just:
```
    # Infinite weights mask
    if max_edge_weight is not None:
        mask = np.logical_and(mask, data <= max_edge_weight)
```
Then, whenever `max_edge_weight is None`, `mask` remains simply the off-diagonal mask, i.e. every edge which was passed by the user.

I still don't fully understand the previous claim that `np.inf` does not get passed to `compute_homology` by the current code.

Reply:
Thanks @gtauzin, I was unnecessarily confused about this. My suggestion is now the following:

1) Remove the if-else block defining `_max_edge_weight`;
2) In the code for `extract_weighted_graph`, replace
```    
    # Infinite weights mask
    if max_edge_weight is not None:
        mask = np.logical_and(mask, data <= max_edge_weight)
```
by
```
    # Infinite weights mask
    if np.issubdtype(adjacency_matrix.dtype, np.float_):
        if (max_edge_weight is None) or np.isposinf(max_edge_weight):
            mask = np.logical_and(mask, np.isfinite(data))
    elif max_edge_weight is not None:
        mask = np.logical_and(mask, data <= max_edge_weight)
```

Reply:
Agreed.